Develop

Sweethezus opened the pull request.

Sweethezus committed:
Initial Player Classes

Initial commit of Player classes. This is how we'll interact with the different players.

Sweethezus committed:
Merge branch 'master' into BrianBranch

Sweethezus committed:
Initial Add of Classes

First pass at adding classes from design document.

Sweethezus committed:
Continued Method Stubbing

Stubbed out more methods in Board class. Added necessary accessors to Location, Piece and Weapon.

greenca6 committed:
Merge branch 'master' into BrianBranch

greenca6 committed:
Rename package to clueless. Create packages according to design. Create stub files. Fix formatting

greenca6 committed:
Rename service impl file.

Sweethezus committed:
Merge pull request #8 from greenca6/BrianBranch

Brian branch

greenca6 committed:
Add initial UI structure, dependencies

Sweethezus committed:
Changes for Compilation

Changed method calls and functions to ensure that the code will compile

Sweethezus committed:
Board Class Changes

All methods and functions within board class have been completed or stubbed out. Need to finalize constructor.

Sweethezus committed:
Player Class

Fully fleshed out player class.

Sweethezus committed:
Fixed testing

Fixed test script to allow tests to run

greenca6 commented on api/src/main/java/com/debugmafia/clueless/model/Board.java:
```diff
@@ -0,0 +1,220 @@
+import java.util.stream.Stream;
+
+public class Board {
+  private BoardLocation[][] grid;
```
After the professor gave feedback, I think this should be a 1 dimensional array. No need for a grid really

Sweethezus replied:
If we are moving away from a 2D structure to a 1D structure, I think a List would be more appropriate. That said, I think we need to be re-think how we search for locations with a one dimensional structure. We'd need to derive a formula that will give us valid locations above and below as well as left and right of our current space, all in one dimensional space.

greenca6 replied:
I think if it were 1D, the algorithm would go like:

1. Iterate over the list, find the location in question (the one we need to find adjacent Locations of)
2. Once found, get the `adjacentTo` Locations, which are stored as class properties
3. Also get the optional `secretPassageTo` Location if it has it
4. Return the list of Locations we got from 2 and 3

greenca6 commented on api/src/main/java/com/debugmafia/clueless/model/Board.java:
```diff
@@ -0,0 +1,220 @@
+  public Set<Card> drawWinningCards() {
+    Set<Card> cardsToReturn;
+
+    if (winningCards == null) {
```
No need for a null check here - just create the `HashSet` in the same line you instantiate it. Also, `HashSet<>` is good enough here, the `Card` is implicit.

greenca6 commented on api/src/main/java/com/debugmafia/clueless/model/Board.java:
```diff
@@ -0,0 +1,220 @@
+
+    if (winningCards == null) {
+      cardsToReturn = new HashSet<Card>();
+      if (!hasDeckBeenShuffled) {
```
I think we want to fail if the deck hasn't been shuffled? This method I don't think should secretly shuffle the deck if it hasn't been shuffled. When a caller uses this method, it's assumed the deck has already been shuffled.

greenca6 replied:
Actually, I think I'm wrong about throwing an error if we haven't shuffled the deck. I think we don't need to worry about that, since we know that `shuffleDeck()` gets called in the constructor. By definition, the caller couldn't have gotten to this code if the deck hasn't been shuffled.

greenca6 commented on api/src/main/java/com/debugmafia/clueless/model/Board.java:
```diff
@@ -0,0 +1,220 @@
+
+  public Set<Card> draw(int numToDraw) {
+
+    HashSet<Card> setToReturn = null;
```
`Set` is the interface, `HashSet` is the implementation. I think this should be:
```java
Set<Card> cards = new HashSet<>();
```

greenca6 commented on api/src/main/java/com/debugmafia/clueless/model/Board.java:
```diff
@@ -0,0 +1,220 @@
+
+    HashSet<Card> setToReturn = null;
+
+    if (hasDeckBeenShuffled && deckOCards.size() >= numToDraw) {
```
Again I think we want to fail here if the deck hasn't been shuffled. It's assumed it has been shuffled by the time someone has called this method.

greenca6 commented on api/src/main/java/com/debugmafia/clueless/model/Board.java:
```diff
@@ -0,0 +1,220 @@
+      Iterator<Card> iterator = deckOCards.iterator();
+
+      while(numToDraw > 0 && iterator.hasNext())
+      {
```
I'm fine with `{` on a new line or on the same line, lets keep it consistent though.

greenca6 commented on api/src/main/java/com/debugmafia/clueless/model/Board.java:
```diff
@@ -0,0 +1,220 @@
+
+    for (BoardLocation[] row : grid) {
+      for(BoardLocation column: row)
+      {
```
Same comment regarding open braces.

greenca6 commented on api/src/main/java/com/debugmafia/clueless/model/Board.java:
```diff
@@ -0,0 +1,220 @@
+    return setToReturn;
+  }
+
+  public Set<BoardLocation> getOpenAdjacentLocations(BoardLocation locationToQuery) {
```
Same comment regarding the professors feedback of our `BoardLocation` class. That class should hold a link to another `BoardLocation`(s) that it is connected to. Which means that this method will need changing after that has been added to that class.

greenca6 commented on api/src/main/java/com/debugmafia/clueless/model/Board.java:
```diff
@@ -0,0 +1,220 @@
+          break;
+        }
+      }
+    }
```
Might be able to condense this logic using Java 8's streaming feature (`findFirst` and/or `filter` would probably do the trick)

greenca6 commented on api/src/main/java/com/debugmafia/clueless/model/Board.java:
```diff
@@ -0,0 +1,220 @@
+
+    if(localLocation == null || localPiece == null)
+    {
+      return null;
```
If this statement executes, something bad has happened. Either the Location they're trying to move to is null, or the piece is. In either case, we should probably throw an error.

greenca6 commented on api/src/main/java/com/debugmafia/clueless/model/Board.java:
```diff
@@ -0,0 +1,220 @@
+    HashSet<Piece> localPieces = new HashSet<Piece>(localLocation.getPieces());
+
+    localPieces.add(localPiece);
+    localLocation.setPieces(localPieces);
```
After we add the piece to the new location we have to do something else - _remove_ the piece from the old location since it'll still be there.

greenca6 commented on api/src/main/java/com/debugmafia/clueless/model/Board.java:
```diff
@@ -0,0 +1,220 @@
+    return this;
+  }
+
+  public Board moveWeapons(Weapon w, BoardLocation to) {
```
Same exact comments as the `movePiece` method regarding using Java 8's streams, error throwing, and removing the weapon from the old location.

greenca6 commented on api/src/main/java/com/debugmafia/clueless/model/Board.java:
```diff
@@ -0,0 +1,220 @@
+
+  }
+
+  public Card getAssociatedCard(Weapon w) {
```
Logic is sound here. I think we can improve it by:

1. Overwriting the `equals()` method in the `Weapon` class, putting the "equals" logic in there that you have here (type are equal, names are equal)
2. Updating this `filter` call to us the `.equals()` method, making this read much cleaner
3. Using the `findFirst` stream method that Java 8 comes with, that way you don't have to do a size check, and return the item at the 0th index

greenca6 replied:
Actually, I'm wrong here about the `.equals` stuff. This is grabbing a card from a weapon/piece/location....so what I said doesn't apply. Oops.

greenca6 commented on api/src/main/java/com/debugmafia/clueless/model/Board.java:
```diff
@@ -0,0 +1,220 @@
+    }
+  }
+
+  public Card getAssociatedCard(Piece p) {
```
Same exact comments as the other `getAssociatedCard` method.

greenca6 commented on api/src/main/java/com/debugmafia/clueless/model/Board.java:
```diff
@@ -0,0 +1,220 @@
+    }
+  }
+
+  public Card getAssociatedCard(BoardLocation l) {
```
Same exact comments as the other `getAssociatedCard` method.

greenca6 commented on api/src/main/java/com/debugmafia/clueless/model/BoardLocation.java:
```diff
@@ -0,0 +1,69 @@
+
+public class BoardLocation {
+  private int xCoord;
+  private int yCoord;
```
See comments regarding the updates to this class based off of the Professors feedback

greenca6 commented on api/src/main/java/com/debugmafia/clueless/model/BoardLocationType.java:
```diff
@@ -0,0 +1,6 @@
+
+  enum BoardLocationType {
+    ROOM, HALLWAY
+  };
```
I think this has to be `public enum BoardLocationType`? Otherwise callers won't be able to access this enum (it's private to this scope). Also the padding to the left can be removed.

greenca6 commented on api/src/main/java/com/debugmafia/clueless/model/Card.java:
```diff
@@ -0,0 +1,24 @@
+    return this.name;
+  }
+
+}
```
Add the `@Override` for the `equals()` method.

greenca6 commented on api/src/main/java/com/debugmafia/clueless/model/CardType.java:
```diff
@@ -0,0 +1,6 @@
+
+  enum CardType {
+    ROOM, WEAPON, PIECE
+  };
```
Same comment as `BoardLocationType`

greenca6 commented on api/src/main/java/com/debugmafia/clueless/model/Piece.java:
```diff
@@ -0,0 +1,21 @@
+  public UUID getUuid() {
+    return this.uuid;
+  }
+}
```
Add the `@Override` for the `equals()` method.

greenca6 commented on api/src/main/java/com/debugmafia/clueless/model/Player.java:
```diff
@@ -0,0 +1,52 @@
+  private String userName;
+  private int playerNum;
+  private Piece character;
+  private boolean isInControl;
```
Don't think we need the `isInControl` flag - the `Game` instance knows this information only.

greenca6 commented on api/src/main/java/com/debugmafia/clueless/model/Player.java:
```diff
@@ -0,0 +1,52 @@
+public class Player {
+
+  private String userName;
+  private int playerNum;
```
Is `playerNum` used for determining turn order?

Sweethezus replied:
More so for who is player 1, to be able to start the game. That was my thought anyway.

greenca6 commented on api/src/main/java/com/debugmafia/clueless/model/Player.java:
```diff
@@ -0,0 +1,52 @@
+  private Set<Card> cardsInHand;
+
+  public Player() {
+  }
```
I think we can remove this constructor. You shouldn't be able to create a `Player` if you don't give a `userName` and `Piece` at a minimum.

greenca6 commented on api/src/main/java/com/debugmafia/clueless/model/Weapon.java:
```diff
@@ -0,0 +1,21 @@
+  public UUID getUuid() {
+    return this.uuid;
+  }
+}
```
Add the `@Override` for the `equals()` method.

Sweethezus replied:
I have a rough understanding of what needs to be done there, but I think you should perform the implementation.

greenca6 replied:
Will do

greenca6 commented on api/src/test/java/com/debugmafia/demo/DemoApplicationTests.java:
```diff
@@ -1,4 +1,4 @@
-package com.debugmafia.demo;
+package com.debugmafia.clueless;
```
This package is correct - but this file should be moved from:

```
api/src/test/java/com/debugmafia/demo
```
to:

```
api/src/test/java/com/debugmafia/clueless
```

Directory structure should match the source code (everything under `api/src/main/java`)

greenca6 reviewed (changes requested):
Some changes to make here. Overall looks about right.

Sweethezus committed:
Review Changes

Implemented changes based on code review

greenca6 commented on api/src/main/java/com/debugmafia/clueless/model/Board.java:
```diff
@@ -0,0 +1,220 @@
+  private Set<Card> deckOCards;
+  private Set<Piece> gamePieces;
+  private Set<Weapon> gameWeapons;
+  private Boolean hasDeckBeenShuffled;
```
Don't think we need this flag on second thought.

greenca6 committed:
Add test stubs. Add equals methods. Slightly restructure Board and BoardLocation

greenca6 committed:
Add Move action class

Sweethezus committed:
Added Actions

Added all the action classes.

emtuls committed:
minor change to method name in Board.java. Added some template stuff to Game.java, will continue working it (needs hard review for sure)

greenca6 committed:
Refactor BoardLocation and Board classes. Reorganize methods.

greenca6 committed:
Remove unused import

greenca6 committed:
Add factory clases for initialization code for locations, weapons, pieces, cards.

Commit:
merge and updated Game.java

Commit:
missing a ton of 'this' keywords

greenca6 committed:
Add setup method for tests

Sweethezus committed:
Cleaned up compilation errors

greenca6 committed:
Merge branch 'develop' into feature/state

greenca6 committed:
Add caching to PieceFactory

greenca6 committed:
Implement ActionType, Lobby, Turn and TurnState

greenca6 committed:
Update Game class to reflect Turn class. Formatting, cleanup.

greenca6 committed:
Update gradle file

greenca6 committed:
Add BoardTests class for testing

greenca6 committed:
Add Lobby methods

greenca6 committed:
Add test files/functions. Add getters in game class

emtuls committed:
updated Game.Java with turnState and gameState Stuff and completed gameState Enum

emtuls committed:
fixed Enum issues

emtuls committed:
yeeaaaa, now it should be fixed

emtuls committed:
ugh, keep forgetting the setters

emtuls committed:
fixed spelling error of accusation in turn.java

Sweethezus committed:
Initial Controller Commit

Initial controller commit. This probably won't compile

greenca6 committed:
Merge develop

greenca6 committed:
Add service classes. Add higher order components. Add basic routing structure.

Sweethezus committed:
Add GameService

Added GameService as well as the implementation, GameServiceImpl. This will serve as the basis for the server.

Sweethezus committed:
Controllers Implemented

Finished first pass at all the controllers. All code does compile.

Sweethezus committed:
Addressed Review Comments

Addressed comments from PR review

Sweethezus committed:
Game update

Updated function calls and methods within Game to near completion

emtuls committed:
added setter method for setting available locations and finished endTurn method in Game.Java

emtuls committed:
mostly finished Game.java, needs look over

Sweethezus committed:
Updated code based on review

Addressed comments made by Mr. Green during review.

Sweethezus committed:
Location check

Added check to make sure piece is in right location before suggestion is allowed

Sweethezus committed:
Merge pull request #14 from greenca6/feature/controllers

Feature/controllers

Sweethezus committed:
Merge branch 'develop' into feature/state

Sweethezus committed:
Updated Game logic

Updated the logic in Game so that we have a list of all players, primarily used for rebuttal purposes. When a player is removed, they are no longer in the activePlayers, but will still be in the allPlayers to be able to refute a claim.

Sweethezus committed:
Fix compilation

Fixed silly compilation mistake

Sweethezus committed:
Merge pull request #13 from greenca6/feature/state

Feature/state

greenca6 committed:
Merge branch 'develop' of https://github.com/greenca6/debug-mafia-demo into feature/initial-ui-structure

Sweethezus commented on api/src/main/java/com/debugmafia/clueless/model/Board.java:
```diff
@@ -0,0 +1,138 @@
+    Piece targetPiece = this.pieces.stream().filter(p -> p.equals(piece)).findFirst().get();
+    Optional<BoardLocation> oldLocation = this.grid.stream().filter(l -> l.containsPiece(targetPiece)).findFirst();
+
+    if (oldLocation.isPresent()) {
```
Shouldn't the piece always be present? Can we just short circuit this logic? Same thing with the weapons.

Sweethezus commented on api/src/main/java/com/debugmafia/clueless/model/BoardLocationFactory.java:
```diff
@@ -0,0 +1,134 @@
+    BoardLocation conservatoryToBallRoom = new BoardLocation(BoardLocationType.HALLWAY, "conservatory to ball room");
+    BoardLocation ballRoomToKitchen = new BoardLocation(BoardLocationType.HALLWAY, "ball room to kitchen");
+
+    /* Connect Rooms and Hallways */
```
I'm not seeing the addition of secret passages anywhere in here

greenca6 replied:
It's a [few lines up](https://github.com/greenca6/debug-mafia-demo/pull/11/files/23b5f6bd8761c7937619951ef717e419e6b68ac9#diff-5c6276eb62c3c79ad6760f31f6225dcfR21)

Sweethezus commented on api/src/main/java/com/debugmafia/clueless/model/CardFactory.java:
```diff
@@ -0,0 +1,39 @@
+    Set<Card> deck = new HashSet<>();
+
+    // Piece Cards
+    deck.add(new Card(CardType.PIECE, "miss scarlet"));
```
Does it make more sense to have the character names be an enumerated type since we have to iterate over them in the Board to sort players by turn order?

greenca6 replied:
Hindsight is 50/50. I think this would make more sense yeah. But, here we are....

greenca6 committed:
completed lobby UI and integration with backend

greenca6 committed:
UI API Integration

Commit:
Merge pull request #10 from greenca6/feature/initial-ui-structure

Add initial UI structure, dependencies

greenca6 committed:
Add force exit button for easier restarts

greenca6 committed:
remove log statements

greenca6 committed:
remove more log statements

Sweethezus committed:
Logic updates

Updated logic so that the next player in the circle will be the first to be able to make a rebuttal. Also changed logic so that the weapons are correctly removed from old locations.

greenca6 commented on api/src/main/java/com/debugmafia/clueless/model/BoardLocation.java:
```diff
@@ -44,7 +44,7 @@ public boolean containsPiece(Piece piece) {
 
   public boolean containsWeapon(Weapon weapon) {
-    return this.weapons.size() > 1 && this.weapons.stream().anyMatch(w -> w.equals(weapon));
+    return this.weapons.size() > 0 && this.weapons.stream().anyMatch(w -> w.equals(weapon));
```
oof, don't know why I did `> 1`

greenca6 commented on api/src/main/java/com/debugmafia/clueless/state/Game.java:
```diff
@@ -28,7 +28,6 @@
   public Game(Set<Player> players) {
     this.board = new Board();
     this.gameState = GameState.IN_PROGRESS;
-    this.players = new ArrayList<>(players);
```
I saw that you moved this line, which is okay and is working fine from what I can tell. Just noting here that I had put this line here originally because the app was crashing when a new Game was created. This was because the `sortAndAssignPlayerList()` method was trying to use `this.players` before it had been assigned, so assigning it here fixed that issue.

Sweethezus replied:
Since this is now being done within sortAndAssignPlayerList, I think we're okay. I had no issues when I was running last night.

greenca6 committed:
Format document
